web/src/main/java/com/epam/esm/controller/CertificateRESTController.java

```diff
@@ -0,0 +1,139 @@
+    @ResponseStatus(HttpStatus.CREATED)
+    public Certificate addCertificate(@RequestBody Certificate certificate) {
+        service.saveEntity(certificate);
+        return certificate;
+    }
+
+
+    /**
+     * @param page the page
+     * @param size the size
+     * @return list certificate
+     */
+    @GetMapping("/certificates")
+    @ResponseStatus(HttpStatus.OK)
+    public CollectionModel<ModelCertificate> listAllCertificates(@RequestParam("page") int page,
+                                                                 @RequestParam("size") int size) {
+        int offset = Pagination.offset(page, size);
+        int totalRecords = service.countAllCertificates();
+        int pages = Pagination.findPages(totalRecords, size);
+        int lastPage = Pagination.findLastPage(pages, size, totalRecords);
+        Link prevLink = linkTo(methodOn(CertificateRESTController.class).listAllCertificates(Pagination.findPrevPage(page), size))
+                .withRel("prev");
+        Link nextLink = linkTo(methodOn(CertificateRESTController.class).listAllCertificates(Pagination.findNextPage(page, lastPage), size))
+                .withRel("next");
+        List<ModelCertificate> models = service.getAllEntity(size, offset);
+        return CollectionModel.of(models, prevLink, nextLink);
+    }
+
+    /**
+     * get certificate by id
+     *
+     * @param id the id
+     * @return certificate
+     */
+    @GetMapping("/certificates/{id}")
+    @ResponseStatus(HttpStatus.OK)
+    public Optional<ModelCertificate> getCertificateById(@PathVariable long id) {
+        Optional<ModelCertificate> model = Optional.ofNullable(service.getEntity(id)).get();
+        model.get().add(linkTo(methodOn(CertificateRESTController.class)
+                .listAllCertificates(1,5))
+                .withRel("certificate")
+                .withType(HttpMethod.GET.name()));;
+        return model;
+    }
+
+    /**
+     * update certificate
+     * enter the fields that you want to update
+     * date not updated except the date
+     *
+     * @param certificate the certificate
+     * @param id          the id
+     * @return updated certificate
+     */
+    @PatchMapping("/certificates/{id}")
+    @ResponseStatus(HttpStatus.OK)
+    public Certificate updateCertificate(@RequestBody OnlyCertificate certificate, @PathVariable long id) {
+        service.updateEntity(id, certificate);
+        return certificateMapper.mapFrom(certificate);
+    }
+
+    /**
+     * delete certificate by id
+     *
+     * @param id the id
+     * @return string response
+     */
+    @DeleteMapping("/certificates/{id}")
+    @ResponseStatus(HttpStatus.OK)
+    public String deleteCertificate(@PathVariable long id) {
+        service.deleteEntity(id);
+        return "Certificate with ID = " + id + " was deleted.";
+    }
+
+    /**
+     * @param name the name
+     * @return certificate by name or part of name
+     */
+    @GetMapping("/certificates/name/{name}")
```

Comment:
This url doesn't folllow REST.
REST use resources.
Thus,
1. /certificates - should return sertificates - that's ok.
2. /name - should return resources "name" in certificates. - that's not ok. Clients don't know from which certificate they should recieve name resources.

My recommendation is change endpoint to "certificates?name={name}" and read this material https://github.com/mjc-school/MJC-School/blob/main/stage%20%233/module%20%234.%20REST%20Basics%20-%20Spring%20Web/lesson%20%231.%20REST%20Basics/content.md

Reply:
The same is for rest of endpoints

Reply:
fixed it, it seems